File: android/src/main/kotlin/com/addcn/lgpl_ffmpeg_flutter/LgplFfmpegFlutterPlugin.kt
package com.addcn.lgpl_ffmpeg_flutter

import android.content.Context
import org.json.JSONObject
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** Controlled Flutter wrapper for an LGPL FFmpeg backend. */
class LgplFfmpegFlutterPlugin :
    FlutterPlugin,
    MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var applicationContext: Context

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "lgpl_ffmpeg_flutter")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result
    ) {
        when (call.method) {
            "readInfo" -> handleReadInfo(call, result)
            "generateCover" -> handleGenerateCover(call, result)
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    private fun handleGenerateCover(
        call: MethodCall,
        result: Result
    ) {
        val videoPath = call.argument<String>("videoPath")
        if (videoPath.isNullOrBlank()) {
            result.error("invalidPath", "videoPath must not be empty.", null)
            return
        }

        val maxLongEdge = call.argument<Int>("maxLongEdge") ?: 1920
        val quality = call.argument<Int>("quality") ?: 95
        if (maxLongEdge <= 0 || quality !in 1..100) {
            result.error(
                "invalidArgument",
                "maxLongEdge must be greater than 0 and quality must be between 1 and 100.",
                null
            )
            return
        }

        runBackend(result) {
            nativeGenerateCover(
                videoPath,
                call.argument<List<Number>>("preferredTimesMs")?.map { it.toLong() }?.toLongArray(),
                maxLongEdge,
                quality,
                applicationContext.cacheDir.absolutePath
            )
        }
    }

    private fun handleReadInfo(
        call: MethodCall,
        result: Result
    ) {
        val videoPath = call.argument<String>("videoPath")
        if (videoPath.isNullOrBlank()) {
            result.error("invalidPath", "videoPath must not be empty.", null)
            return
        }

        runBackend(result) { nativeReadInfo(videoPath) }
    }

    private fun runBackend(
        result: Result,
        block: () -> String
    ) {
        if (!backendLoaded) {
            result.error(
                "ffmpegUnavailable",
                backendLoadError ?: "LGPL FFmpeg dynamic libraries are not bundled.",
                null
            )
            return
        }

        try {
            val json = JSONObject(block())
            val errorCode = json.optString("errorCode", "")
            if (errorCode.isNotEmpty()) {
                result.error(
                    errorCode,
                    json.optString("message", errorCode),
                    null
                )
                return
            }
            result.success(jsonToMap(json))
        } catch (exception: NativeVideoProcessException) {
            result.error(exception.code, exception.message, exception.details)
        } catch (exception: Throwable) {
            result.error("unknown", exception.message ?: "Unexpected native error.", null)
        }
    }

    private fun jsonToMap(json: JSONObject): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>()
        val keys = json.keys()
        while (keys.hasNext()) {
            val key = keys.next()
            map[key] = if (json.isNull(key)) null else json.get(key)
        }
        return map
    }

    private external fun nativeReadInfo(videoPath: String): String

    private external fun nativeGenerateCover(
        videoPath: String,
        preferredTimesMs: LongArray?,
        maxLongEdge: Int,
        quality: Int,
        cacheDir: String
    ): String

    private class NativeVideoProcessException(
        val code: String,
        override val message: String,
        val details: Any? = null
    ) : RuntimeException(message)

    companion object {
        private val backendLoadError: String?
        private val backendLoaded: Boolean

        init {
            var loaded = false
            var error: String? = null
            try {
                System.loadLibrary("avutil")
                System.loadLibrary("swresample")
                System.loadLibrary("swscale")
                System.loadLibrary("avcodec")
                System.loadLibrary("avformat")
                System.loadLibrary("lgpl_ffmpeg_flutter_backend")
                loaded = true
            } catch (throwable: Throwable) {
                error = throwable.message
            }
            backendLoaded = loaded
            backendLoadError = error
        }
    }
}

File: ios/Classes/LgplFfmpegFlutterPlugin.swift
import Flutter
import UIKit
import lgpl_ffmpeg_flutter

public class LgplFfmpegFlutterPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "lgpl_ffmpeg_flutter", binaryMessenger: registrar.messenger())
    let instance = LgplFfmpegFlutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "readInfo":
      handleReadInfo(call, result: result)
    case "backendInfo":
      result(mapResult(from: lgpl_ffmpeg_flutter_backend_info()))
    case "generateCover":
      handleGenerateCover(call, result: result)
    case "extractFrame":
      handleExtractFrame(call, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func handleReadInfo(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let arguments = call.arguments as? [String: Any],
          let videoPath = arguments["videoPath"] as? String,
          !videoPath.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
      result(FlutterError(code: "invalidPath", message: "videoPath must not be empty.", details: nil))
      return
    }

    result(mapResult(from: lgpl_ffmpeg_flutter_read_info(videoPath)))
  }

  private func handleGenerateCover(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let arguments = call.arguments as? [String: Any],
          let videoPath = arguments["videoPath"] as? String,
          !videoPath.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
      result(FlutterError(code: "invalidPath", message: "videoPath must not be empty.", details: nil))
      return
    }

    let maxLongEdge = arguments["maxLongEdge"] as? Int ?? 1920
    let quality = arguments["quality"] as? Int ?? 95
    guard maxLongEdge > 0, (1...100).contains(quality) else {
      result(FlutterError(
        code: "invalidArgument",
        message: "maxLongEdge must be greater than 0 and quality must be between 1 and 100.",
        details: nil
      ))
      return
    }

    let preferredTimesMs = arguments["preferredTimesMs"] as? [Int64] ?? []
    let cacheDir = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)
      .first?
      .path ?? NSTemporaryDirectory()
    let nativeResult = preferredTimesMs.withUnsafeBufferPointer { buffer in
      lgpl_ffmpeg_flutter_generate_cover(
        videoPath,
        buffer.baseAddress,
        Int32(buffer.count),
        Int32(maxLongEdge),
        Int32(quality),
        cacheDir
      )
    }
    result(mapResult(from: nativeResult))
  }

  private func handleExtractFrame(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let arguments = call.arguments as? [String: Any],
          let videoPath = arguments["videoPath"] as? String,
          !videoPath.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
      result(FlutterError(code: "invalidPath", message: "videoPath must not be empty.", details: nil))
      return
    }

    let maxLongEdge = arguments["maxLongEdge"] as? Int ?? 1920
    let quality = arguments["quality"] as? Int ?? 95
    guard maxLongEdge > 0, (1...100).contains(quality) else {
      result(FlutterError(
        code: "invalidArgument",
        message: "maxLongEdge must be greater than 0 and quality must be between 1 and 100.",
        details: nil
      ))
      return
    }

    let timesMs = [arguments["timeMs"] as? Int64 ?? 0]
    let cacheDir = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)
      .first?
      .path ?? NSTemporaryDirectory()
    let nativeResult = timesMs.withUnsafeBufferPointer { buffer in
      lgpl_ffmpeg_flutter_generate_cover(
        videoPath,
        buffer.baseAddress,
        Int32(buffer.count),
        Int32(maxLongEdge),
        Int32(quality),
        cacheDir
      )
    }
    result(mapResult(from: nativeResult))
  }

  private func mapResult(from nativeResult: UnsafeMutablePointer<CChar>?) -> Any {
    guard let nativeResult else {
      return FlutterError(
        code: "ffmpegUnavailable",
        message: "LGPL FFmpeg backend did not return a result.",
        details: nil
      )
    }
    let jsonText = String(cString: nativeResult)
    lgpl_ffmpeg_flutter_free_string(nativeResult)

    guard let data = jsonText.data(using: .utf8),
          let object = try? JSONSerialization.jsonObject(with: data),
          let map = object as? [String: Any] else {
      return FlutterError(code: "unknown", message: "Invalid backend JSON.", details: nil)
    }
    if let errorCode = map["errorCode"] as? String, !errorCode.isEmpty {
      return FlutterError(
        code: errorCode,
        message: map["message"] as? String ?? errorCode,
        details: nil
      )
    }
    return map
  }
}
